hw/bsp/dialog_da1469x-dk-pro/otp_tool.py

```diff
@@ -0,0 +1,538 @@
+    cmd: int
+    segment: int
+    slot: int
+
+
+class cmd_append_value(NamedTuple):
+    som: int
+    cmd: int
+    length: int
+
+
+class cmd_response(NamedTuple):
+    som: int
+    cmd: int
+    status: int
+    length: int
+
+
+class cmd_flash_op(NamedTuple):
+    som: int
+    cmd: int
+    address: int
+    length: int
+
+
+def crc16(data: bytearray, offset, length):
+    if data is None or offset < 0:
+        return
+
+    if offset > len(data) - 1 and offset + length > len(data):
+        return 0
+
+    crc = 0
+    for i in range(length):
+        crc ^= data[offset + i] << 8
+
+        for j in range(8):
+            if (crc & 0x8000) > 0:
+                crc = (crc << 1) ^ 0x1021
+            else:
+                crc = crc << 1
+    return crc & 0xFFFF
+
+
+def validate_slot_index(ctx, param, value):
+    if value in range(8):
+        return value
+    else:
+        raise click.BadParameter("Slot value has to be between 0 and 7")
+
+
+@click.option('-i', '--index', required=True, type=int, help='key slot index',
+              callback=validate_slot_index)
+@click.option('-u', '--uart', required=True, help='uart port')
+@click.option('-s', '--segment', type=click.Choice(['signature', 'data',
+              'qspi']), help='OTP segment', required=True)
+@click.command(help='Read a single key from OTP key segment')
+def otp_read_key(index, segment, uart):
+    seg_map = {'signature': 0, 'data': 1, 'qspi': 2}
+
+    try:
+        ser = serial.Serial(port=uart, baudrate=115200, timeout=1,
+                            bytesize=8, stopbits=serial.STOPBITS_ONE)
+    except serial.SerialException:
+        raise SystemExit("Failed to open serial port")
+
+    cmd = cmd_read_key(0xaa55aa55, Cmd.OTP_READ_KEY, seg_map[segment], index)
+    data = struct.pack('IIII', *cmd)
+
+    try:
+        ser.write(data)
+    except serial.SerialException:
+        raise SystemExit("Failed to write to %s" % uart)
+
+    data = ser.read(16)
+    if len(data) != 16:
+        raise SystemExit("Failed to receive response, exiting")
+
+    response = cmd_response._make(struct.unpack_from('IIII', data))
+    if response.status == 0:
+        key = ser.read(32)
+        print("key:" + key.hex())
+    else:
+        raise SystemExit("Error reading key with status %s" %
+                         hex(response.status))
+
+
+@click.argument('infile')
+@click.option('-u', '--uart', required=True, help='uart port')
+@click.option('-i', '--index', required=True, type=int, help='key slot index',
+              callback=validate_slot_index)
+@click.option('-s', '--segment', type=click.Choice(['signature', 'data',
+              'qspi']), help='OTP segment', required=True,)
+@click.command(help='Write a single key to OTP key segment')
+def otp_write_key(infile, index, segment, uart):
+
+    key = bytearray()
+    try:
+        with open(infile, "rb") as f:
+            buf = f.read(32)
+            key = struct.unpack('IIIIIIII', buf)
+    except IOError:
```

Comment:
I believe something could go wrong here if you correctly open and read a file that is smaller than 32 bytes (`read` won't fail when there is less than 32), but `struct.unpack` is going to raise an exception that is not derived from `IOError`. One option would be to do a catch all (`except Exception`), or handle specific extra exceptions (`struct.error`).

Reply:
good point, let me handle that case.

Reply:
at some point i also need to handle pem vs base64 encoded files.